qt/scientific_interfaces/Indirect/IndirectFunctionBrowser/ConvFunctionModel.cpp

```diff
@@ -0,0 +1,689 @@
+        if (isBackgroundSet) {
+          throw std::runtime_error("Function has wrong structure.");
+        }
+        m_backgroundType = BackgroundType::Linear;
+        isBackgroundSet = true;
+      } else if (name == "Convolution") {
+        checkConvolution(innerFunction);
+      }
+    }
+  }
+  m_model.setFunction(fun);
+}
+
+void ConvFunctionModel::checkConvolution(IFunction_sptr fun) {
+  bool isFitTypeSet = false;
+  bool isResolutionSet = false;
+  for (size_t i = 0; i < fun->nFunctions(); ++i) {
+    auto innerFunction = fun->getFunction(i);
+    auto const name = innerFunction->name();
+    if (name == "Resolution") {
+      if (isResolutionSet) {
+        throw std::runtime_error("Function has wrong structure.");
+      }
+      isResolutionSet = true;
+    } else if (name == "CompositeFunction") {
+      checkComposite(innerFunction);
+    } else if (FitTypeStringToEnum.count(name) == 1) {
+      if (isFitTypeSet) {
+        throw std::runtime_error("Function has wrong structure.");
+      }
+      m_fitType = FitTypeStringToEnum[name];
+      m_isQDependentFunction = FitTypeQDepends[m_fitType];
+      isFitTypeSet = true;
+    } else if (name == "DeltaFunction") {
+      m_hasDeltaFunction = true;
+    } else {
+      clear();
+      throw std::runtime_error("Function has wrong structure.");
+    }
+  }
+} // namespace IDA
+
+void ConvFunctionModel::checkComposite(IFunction_sptr fun) {
+  bool isFitTypeSet = false;
+  for (size_t i = 0; i < fun->nFunctions(); ++i) {
+    auto innerFunction = fun->getFunction(i);
+    auto const name = innerFunction->name();
+    if (name == "Lorentzian") {
+      if (isFitTypeSet && m_fitType != FitType::OneLorentzian) {
+        throw std::runtime_error("Function has wrong structure.");
+      }
+      if (isFitTypeSet)
+        m_fitType = FitType::TwoLorentzians;
+      else
+        m_fitType = FitType::OneLorentzian;
+      m_isQDependentFunction = false;
+      isFitTypeSet = true;
+
+    } else if (FitTypeStringToEnum.count(name) == 1) {
+      if (isFitTypeSet) {
+        throw std::runtime_error("Function has wrong structure.");
+      }
+      m_fitType = FitTypeStringToEnum[name];
+      m_isQDependentFunction = FitTypeQDepends[m_fitType];
+      isFitTypeSet = true;
+    } else if (name == "DeltaFunction") {
+      m_hasDeltaFunction = true;
+    } else {
+      clear();
+      throw std::runtime_error("Function has wrong structure.");
+    }
+  }
+}
+
+IFunction_sptr ConvFunctionModel::getFitFunction() const {
+  return m_model.getFitFunction();
+}
+
+void ConvFunctionModel::setQValues(const std::vector<double> &qValues) {
+  m_qValues = qValues;
+}
+
+FitType ConvFunctionModel::getFitType() const { return m_fitType; }
+BackgroundType ConvFunctionModel::getBackgroundType() const {
+  return m_backgroundType;
+}
+
+bool ConvFunctionModel::hasFunction() const { return m_model.hasFunction(); }
+
+void ConvFunctionModel::addFunction(const QString &prefix,
```

Comment:
Since this is a new model, could we strip all the QStrings out and replace them with std::strings
The presenter can convert on call, then long term you can convert between in the view (to help encapsulate them)

Reply:
Whilst this is a new Model it inherits from FunctionModel.h which uses Q strings throughout it's function declarations so the overrides will not function properly with normal strings. In also acts as an adapter around ConvolutionFunctionModel which also inherits from FunctionModel.h so would need to convert back to QStrings before calling on this class anyway. 

I agree though that there is some refactoring work to be done here both due to the fact that the QObjects have got so far out of the view and the fact that ConvFunctionModel.h and ConvolutionFunctionModel do very similar things whilst one owns the other but this is a larger job than I want to tackle here. I think the original motivation here is that the the view,presenter and Model here are seen as one encapsulated widget between them but from experience debugging working with QStrings and QContainers all over the place is annoying for debugging and code readability.